### ISSUE TITLE
API call to get track playback quality

### ISSUE THREAD

**Issue by fmarzocca** [tehkillerbee/mopidy-tidal#162]
I am looking for a way to get from mopidy-tidal (maybe an API call?) the quality of the playing track (i.e.: HI_RES_LOSSLESS, LOSSLESS, etc), for my extension.
Is it feasible, or should I parse the log?

**Comment by tehkillerbee** [comment]
You cannot get this info through mopidy-tidal, but you can get it indirectly through the tidalapi:
`session.track(track_id).media_metadata_tags`

The list of media_metadata_tags will contain the available quality for your track. If the quality requested by your mopidy-tidal configuration is not available, a comparable quality will be used. So if HIRES_LOSSLESS was requested but only LOSSLESS was available, LOSSLESS will be used for playback.

If the Stream object is accessed instead, more details are available wrt. the audio resolution, codec, quality etc. of the available audio stream for the selected track:
```python
stream = session.track(track_id).get_stream()
...
quality = stream.audio_quality # LOW, HIGH, LOSSLESS, HI_RES, HI_RES_LOSSLESS
bit_depth = stream.bit_depth # bit_depth: int = 16
sample_rate = stream.sample_rate #      sample_rate: int = 44100
```

**Comment by fmarzocca** [comment]
Thank you very much for your info's!

**Comment by fmarzocca** [comment]
One question (waiting for the functionality in mopidy): if I run a separate python script to access tidalapi, it will ask again for login authentication... is it correct? How can I use the same auth I have with mopidy-tidal?

**Comment by tehkillerbee** [comment]
@fmarzocca Not sure what you are trying to do. Does your extension rely on mopidy-tidal? if so, it must use the session/authentication provided by this extension. But you could of course modify mopidy-tidal yourself, and access the tidalapi backend directly.

If you use tidalapi "standalone", you just have to create a new session - or reuse the session json generated by mopidy-tidal. It can also be loaded by tidalapi. A good starting point is one of the examples here: https://github.com/tamland/python-tidal/tree/master/examples

**Comment by fmarzocca** [comment]
> or reuse the session json generated by mopidy-tidal.

Yes, this is the way I am going to take. 
Thank you!

**Comment by fmarzocca** [comment]
@tehkillerbee, I made a standalone script which is now able to extract quality and codec from playing track uri, (I am getting the stream from tidalapi).
I have just a question to make it more generalized: as I have to set the audio quality in the tidalapi session, how can I read the content of the related config line of mopidy-tidal?

**Comment by fmarzocca** [comment]
> how can I read the content of the related config line of mopidy-tidal

If I find the way for this, I can embed the mod in my extension, making it available to all of those who are using `mopidy-tidal`.

<img width="368" alt="Screenshot 2024-03-21 alle 11 18 47" src="https://github.com/tehkillerbee/mopidy-tidal/assets/8530030/f2dca122-4a14-450b-9507-bbe0cae97d59">

**Comment by tehkillerbee** [comment]
> how can I read the content of the related config line of mopidy-tidal

The config params are defined here:
https://github.com/tehkillerbee/mopidy-tidal/blob/e2aa9de2fc8f4cdf66305d8b3f76e4b109043c90/mopidy_tidal/__init__.py#L28

The config params are accessed here.
https://github.com/tehkillerbee/mopidy-tidal/blob/e2aa9de2fc8f4cdf66305d8b3f76e4b109043c90/mopidy_tidal/backend.py#L28

So what you are proposing is essentially an extension to mopidy-tidal? You can also consider if it makes sense to merge it in.

**Comment by fmarzocca** [comment]
> So what you are proposing is essentially an extension to mopidy-tidal?

Not really. My extension enables user to interact with mopidy through MQTT. I'd like it to report Track Quality/Codec too, if the user has mopidy-tidal installed. This while waiting for https://github.com/mopidy/mopidy/issues/2085

**Comment by fmarzocca** [comment]
Yes, getting the tidal quality from mopidy-tidal is still an issue for me.

**Comment by fmarzocca** [comment]
@tehkillerbee 
Johannes, in my script that grab quality/code I had to hardcode my setting:

`session.audio_quality = Quality.hi_res`

Can you give me an hint about how to get this setting form mopidy-tidal settings?

**Comment by fmarzocca** [comment]
@tehkillerbee 
i was thinking that it  would be very useful if `mopidy-tidal `exported a function like this one, so as to make it available externally:
```
def getTrackQuality(track_id):
    trk_details = {}
    stream = session.track(track_id).get_stream()
    manifest = stream.get_stream_manifest()
    trk_details['quality'] = stream.audio_quality
    trk_details['codec'] = manifest.get_codecs()
    trk_details["depth"] = stream.bit_depth
    trk_deatils["rate"] = stream.sample_rate
    return (json.dumps(trk_details))
```

**Comment by tehkillerbee** [comment]
> Can you give me an hint about how to get the quality setting form mopidy-tidal settings, so that I don't hardcode it in my script?

The extension config is read [here](https://github.com/tehkillerbee/mopidy-tidal/blob/e2aa9de2fc8f4cdf66305d8b3f76e4b109043c90/mopidy_tidal/backend.py#L28)  and each field can be read [here](https://github.com/tehkillerbee/mopidy-tidal/blob/e2aa9de2fc8f4cdf66305d8b3f76e4b109043c90/mopidy_tidal/backend.py#L80)

> i was thinking that it would be very useful if mopidy-tidal could export a function like this one, so as to make it available externally.

I think I understand the point of this addition but I am not sure this is the right way to do it, as they would result in extra calls to the API for information that we already have.. Perhaps we should consider extending the TidalPlaybackProvider [here](https://github.com/tehkillerbee/mopidy-tidal/blob/master/mopidy_tidal/playback.py#L40) so it populates the variables when playback begins (instead of just printing them to the log) and then add a function similar to your suggestion that allows returning the quality parameters for the playing track.

**Comment by fmarzocca** [comment]
> Perhaps we should consider extending the TidalPlaybackProvider [here](https://github.com/tehkillerbee/mopidy-tidal/blob/master/mopidy_tidal/playback.py#L40) so

of course I posted those lines in my message just to give you an idea, but I agree that the right place is in TidalPlaybackProvider.
